source/extensions/upstreams/http/tcp/upstream_request.cc

```diff
@@ -87,6 +87,11 @@ void TcpUpstream::resetStream() {
 
 void TcpUpstream::onUpstreamData(Buffer::Instance& data, bool end_stream) {
   upstream_request_->decodeData(data, end_stream);
+  // This ensures that if we get a reset after end_stream we won't propagate two
+  // "end streams" to the upstream_request_.
+  if (end_stream) {
```

Comment:
once we have upstream filters, will it be possible to get an upstream "end stream "before we send headers?  I think it could be the case, so I think we're going to want to do null pointer checks for upstream_request in few other places, WDYT?

Reply:
Yeah good point, I'll go over all the functions